### DISCUSSION TITLE
Create storage bucket using SQL?

### DISCUSSION THREAD

**Discussion by tskytt** [supabase/supabase#3528] [Questions] [answered]
Is this good enough for creating a new bucket?:

`insert into storage.buckets (id, name) values ('avatars', 'avatars')`

Good enough meaning it correctly creates the s3 bucket behind the scenes etc.

Another option would be to enable the http extension and do a POST to /storage/v1/bucket/ in the SQL with the proper payload. However, this does not seem to work on my insert trigger function on auth.users. It generates an error: "http_request" does not exist. Don't know why. It works after that trigger in other functions.

**Comment by GaryAustin1** [comment, marked as answer]
This answer appears to be wrong as to what a bucket is as per below.
No.  They have to be created thru the storage API. Just remember buckets in S3 are limited (100 I think) depending on what your trying to do.

**Reply by tskytt** [reply]
Thanks and good point about the s3 limit of 100 buckets. I guess it applies to supabase storage as well because it uses s3 even though the docs lack this info.

**Reply by grschafer** [reply]
I think the hosted supabase platform uses a single S3 bucket for all tenants and objects are created with an S3 key/path like: `${tenantId}/${bucketName}/${objectName}`. See the relevant code here: https://github.com/supabase/storage-api/blob/f687bfff6f25d89283acf00a7061f7fd8152c832/src/routes/object/createObject.ts#L84

You can see in the [createBucket code](https://github.com/supabase/storage-api/blob/f687bfff6f25d89283acf00a7061f7fd8152c832/src/routes/bucket/createBucket.ts#L41) that it doesn't make a request to S3 at all.

So, I would think that creating a bucket via SQL works, because it's equivalent to what the storage API's create-bucket endpoint does.

**Reply by GaryAustin1** [reply]
Interesting.  Looks that way.  One big s3 bucket for a number of users with a key.  Then "Fake" buckets under each key.  So no limit then.  Not I sure I personally would depend on the tables "owned" by the storage AI (like limited suggeted use of Auth table) but to each his own.   Sorry for the apparent mis-information based on my assumption of what a bucket is.  I can't unmark the answer though.
edit: adding about concern to rely on storage object tables... it appears on 8/17 SB did some sort of change to the tables as part of a migration "ALTER TABLE storage.buckets ADD COLUMN "public" boolean default false;"
